tonic/src/codec/compression.rs

```diff
@@ -10,62 +10,79 @@ pub(crate) const ENCODING_HEADER: &str = "grpc-encoding";
 pub(crate) const ACCEPT_ENCODING_HEADER: &str = "grpc-accept-encoding";
 
 /// Struct used to configure which encodings are enabled on a server or channel.
-#[derive(Debug, Default, Clone, Copy)]
+///
+/// Represents an ordered list of compression encodings that are enabled.
+#[derive(Debug, Clone, Copy)]
 pub struct EnabledCompressionEncodings {
-    #[cfg(feature = "gzip")]
-    pub(crate) gzip: bool,
-    #[cfg(feature = "zstd")]
-    pub(crate) zstd: bool,
+    inner: [Option<CompressionEncoding>; 2],
 }
 
 impl EnabledCompressionEncodings {
-    /// Check if a [`CompressionEncoding`] is enabled.
-    pub fn is_enabled(&self, encoding: CompressionEncoding) -> bool {
-        match encoding {
-            #[cfg(feature = "gzip")]
-            CompressionEncoding::Gzip => self.gzip,
-            #[cfg(feature = "zstd")]
-            CompressionEncoding::Zstd => self.zstd,
+    /// Enable a [`CompressionEncoding`].
+    ///
+    /// Adds the new encoding to the end of the encoding list.
+    pub fn enable(&mut self, encoding: CompressionEncoding) {
+        for e in self.inner.iter_mut() {
+            match e {
+                Some(e) if *e == encoding => return,
+                None => {
+                    *e = Some(encoding);
+                    return;
+                }
+                _ => continue,
+            }
         }
     }
 
-    /// Enable a [`CompressionEncoding`].
-    pub fn enable(&mut self, encoding: CompressionEncoding) {
-        match encoding {
-            #[cfg(feature = "gzip")]
-            CompressionEncoding::Gzip => self.gzip = true,
-            #[cfg(feature = "zstd")]
-            CompressionEncoding::Zstd => self.zstd = true,
+    /// Remove the last [`CompressionEncoding`].
+    pub fn pop(&mut self) -> Option<CompressionEncoding> {
+        for entry in self.inner.iter_mut().rev() {
+            match entry {
+                Some(e) => {
+                    let removed = *e;
+                    *entry = None;
+                    return Some(removed);
+                }
+                _ => continue,
+            }
         }
+
+        None
```

Comment:
This could be written without looping.

```suggestion
        if let Some(entry) = self.inner.iter_mut().rev().find(|entry| entry.is_some()) {
            entry.take()
        } else {
            None
        }
```

Reply:
Changed it to

```rust
self.inner
            .iter_mut()
            .rev()
            .find(|entry| entry.is_some())?
            .take()
```